Userland/Libraries/LibC/pwd.cpp

```diff
@@ -134,123 +128,71 @@ int getpwent_r(struct passwd* passwd_buf, char* buffer, size_t buffer_size, stru
-
-    VERIFY(ok);
-
-    *result = pwd;
-    pwd->pw_name = buf_name;
-    pwd->pw_passwd = buf_passwd;
-    pwd->pw_gecos = buf_gecos;
-    pwd->pw_dir = buf_dir;
-    pwd->pw_shell = buf_shell;
-}
-
-int getpwnam_r(char const* name, struct passwd* pwd, char* buf, size_t buflen, struct passwd** result)
-{
-    // FIXME: This is a HACK!
-    TemporaryChange name_change { s_name, {} };
-    TemporaryChange passwd_change { s_passwd, {} };
-    TemporaryChange gecos_change { s_gecos, {} };
-    TemporaryChange dir_change { s_dir, {} };
-    TemporaryChange shell_change { s_shell, {} };
-
     setpwent();
-    bool found = false;
-    while (auto* pw = getpwent()) {
-        if (!strcmp(pw->pw_name, name)) {
-            found = true;
-            break;
+    for (;;) {
+        if (auto rc = getpwent_r(pwd, buffer, bufsize, result); rc != 0)
+            return rc;
+        if (!strcmp(pwd->pw_name, name)) {
+            *result = pwd;
+            return 0;
         }
     }
 
-    if (!found) {
-        *result = nullptr;
-        return 0;
-    }
-
-    auto const total_buffer_length = s_name.length() + s_passwd.length() + s_gecos.length() + s_dir.length() + s_shell.length() + 5;
-    if (buflen < total_buffer_length)
-        return ERANGE;
-
-    construct_pwd(pwd, buf, result);
-    return 0;
+    *result = nullptr;
+    // NOTE: POSIX doesn't specify which error code to return in this case,
+    //       so we just match what Linux does and return ENOENT.
+    return ENOENT;
 }
 
-int getpwuid_r(uid_t uid, struct passwd* pwd, char* buf, size_t buflen, struct passwd** result)
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid.html
+int getpwuid_r(uid_t uid, struct passwd* pwd, char* buffer, size_t bufsize, struct passwd** result)
 {
-    // FIXME: This is a HACK!
-    TemporaryChange name_change { s_name, {} };
-    TemporaryChange passwd_change { s_passwd, {} };
-    TemporaryChange gecos_change { s_gecos, {} };
-    TemporaryChange dir_change { s_dir, {} };
-    TemporaryChange shell_change { s_shell, {} };
-
     setpwent();
-    bool found = false;
-    while (auto* pw = getpwent()) {
-        if (pw->pw_uid == uid) {
-            found = true;
-            break;
+    for (;;) {
+        if (auto rc = getpwent_r(pwd, buffer, bufsize, result); rc != 0)
+            return rc;
+        if (pwd->pw_uid == uid) {
+            *result = pwd;
+            return 0;
         }
     }
 
-    if (!found) {
-        *result = nullptr;
-        return 0;
-    }
-
-    auto const total_buffer_length = s_name.length() + s_passwd.length() + s_gecos.length() + s_dir.length() + s_shell.length() + 5;
-    if (buflen < total_buffer_length)
-        return ERANGE;
-
-    construct_pwd(pwd, buf, result);
-    return 0;
+    *result = nullptr;
+    // NOTE: POSIX doesn't specify which error code to return in this case,
+    //       so we just match what Linux does and return ENOENT.
+    return ENOENT;
```

Comment:
This code is unreachable because the only ways to break out of the loop are returning.